generic ephemeral inline volumes KEP

pohly opened the pull request:
This describes an alternative, more general approach for specifying volumes inside a pod spec ("inline"), with a lifecycle of the volumes that is tied to the lifecycle of the pod ("ephemeral").

Enhancement issue: https://github.com/kubernetes/enhancements/issues/1698

msau42 commented:
/assign @saad-ali

pohly commented:
During a design review meeting (recording will become available) we agreed to move ahead with investigating this because it looks promising.

We identified the need to clarify where the PVC creation will be handled. Options:
- a webhook (but might not be viable because ownership relationship depends on pod uid, which might not be available yet)
- volume scheduling code in kube-scheduler: doable, but not the right place
- a new controller alongside the existing PV controller in kube-controller-manager: most promising solution

I'll investigate that.

saad-ali commented:
Meeting Notes:

* Admission plugin? Webhook? Or Controller?
  * @jsafrane - pod with new type of volume - admission plugin that sees this volume, creates PVC and replace the volume.
    * @pohly - what happen if that inline spec gets updated? mutable fields like size? reclaim policy? I guess we could make it all immutable by only looking it at creation time and making spec immutable at cluster creation.
  * @jsafrane - API folks say no more new admission plugin, suggest webhook instead
  * @pohly - That's why I'd prefer a volume controller or scheduler.
  * @msau42 - architecturally scheduler doesn't make sense.
  * @jsafrane - Separate controller should create PVCs instead of scheduler.
  * @msau42 - agreed, should live in PV controller or new volume controller
  * TODO: @pohly while investigate where to implement
* SIG Arch is likely to ask if use case is strong enough to change pod spec?
  * @pohly - So many features necessary to add to inline volumes, this would make it easy
* @jsafrane - I don't like that it uses `PersistentVolumeClaim` for something that is not `Persistent`
  * @saad-ali agreed
  * @msau42  - we could do a v2
  * @pohly - not necessary; this is just a case where PVC is known to go away.

pohly commented:
> a webhook (but might not be viable because ownership relationship depends on pod uid, which might not be available yet)

I think we can rule out this option exactly for this reason: while "admission controllers may sometimes have side effects" (https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/), they also need a "corresponding reclamation or reconciliation process" because pod  creation comes later. Not being able to rely on the ownership relation seems complicated.

> a new controller alongside the existing PV controller in kube-controller-manager

Let's focus on that.

pohly commented:
@jsafrane: I've finished updating the document with the `kube-controller-manager` approach. I think it is ready for review now.

If I remember correctly, we volunteered you as reviewer because you said that you like this approach, right? :stuck_out_tongue: Can you have a look?

jsafrane commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,540 @@
+
+## Design Details
+
+### PVC meta data
```
Please mention somewhere that the newly created PVC is subject of namespace quota as any other PVC and does not allow user cheating.

pohly replied:
Probably belongs under risks and mitigations. I've added something there.

jsafrane commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,540 @@
+
+## Design Details
+
+### PVC meta data
```
Related to above, please note that the new controller reports any errors on PVC creation as events on the Pod. This way, user can get for example quota errors.

pohly replied:
Added under implementation details, together with name collision.

jsafrane commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
````diff
@@ -0,0 +1,540 @@
+A new volume source will be introduced:
+
+```
+typedef InlineVolumeSource struct {
````
Naming is hard. How is this `InlineVolumeSource` different from other inline volume sources (say `ISCSIVolumeSource`)? The name should imply that it is ephemeral (won't persist data beyond Pod lifetime)... `EphemeralVolume`?

pohly replied:
Including "ephemeral" in the name forever rules out the possibility to make the volumes persistent. I don't think it'll be useful, but I'd prefer to not paint us into a corner here.

The more important aspect is that this inline volume uses the claim mechanism and normal provisioning. How about `ClaimVolumeSource`? It is more accurate, but doesn't role of the tongue that easily.

pohly replied:
`VolumeClaimVolumeSource` is another option, albeit with `Volume` in the name twice.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
````diff
@@ -0,0 +1,540 @@
+
+```
+typedef InlineVolumeSource struct {
+    VolumeClaimTemplate PersistentVolumeClaim
````
The current approach follows how volumes are created in the stateful set. But yes, let's discuss this during the API review.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,540 @@
+
+When downgrading to a cluster which does not support generic inline
+volumes (either by disabling the feature flag or an actual version
+downgrade), pods using generic inline volumes will no longer be
+scheduled or started.
```
I've removed the "be scheduled".

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
````diff
@@ -0,0 +1,540 @@
+```
+typedef InlineVolumeSource struct {
+    VolumeClaimTemplate PersistentVolumeClaim
+    ReadOnly            bool
````
The driver itself or a mechanism like the image populator might create the volume with data, which then might be supposed to be left unmodified. Cloning is indeed another example.

But I'm making this up on-the-fly :grin: The original motivation was to have feature-parity with `PersistentVolumeClaimVolumeSource`, without having considered all implications.

It's definitely useful to write this down, so let me add a use-case.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,540 @@
+through the feature gate that will be added for the feature.
+
+Alternatively, a label on a namespace could be used to disable the
+feature just for that namespace.
```
Good idea, I hadn't thought of that.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,540 @@
+The `GenericInlineVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `InlineVolumeSource`,
+- kubelet and scheduler accept such volumes.
```
Code that merely copies a `Pod.Spec.Volumes` IMHO shouldn't need changes: if it gets an `InlineVolumeSource`, it can pass that through and let those components who really have to do something special with it check the feature gate.

Let's check during the implementation phase what changes are needed where. For now I rephrased that line item a bit to be more open-ended.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,540 @@
+
+### Late binding
+
+Ideally, provisioning should use late binding (aka
+`WaitForFirstConsumer`). The initial implementation assumes that the
+user is taking care of that by selecting that provisioning mode in the
+storage class for the PVC.
+
+Later, `kube-scheduler` and `external-provisioner` can be changed to
+automatically enable late binding for PVCs which are owned by a pod.
```
I couldn't find a reason why late binding should *not* be used. The volume is clearly going to be created for use by this one pod, so using late binding seems like the right thing to do (don't create volumes unless the pod can start, better scheduling) in all cases.

By enabling late binding for inline volumes, we reduce the amount of storage classes which have to be defined in the cluster. It might not be relevant, though, if all classes already have late binding enabled.

jsafrane commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,561 @@
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  Yes, by disabling the feature gates. Existing pods will continue to run and
+  volumes will be deleted once they stop. Existing pods with generic inline
```
> Existing pods will continue to run and  volumes will be deleted once they stop.

If someone just restarts kubelet with the feature gate disabled, it won't be able to unmount InLine volumes from running Pods. I haven't found documentation how to actually flip a feature gate in a running cluster. Can you please add a note that node must be drained before flipping the feature gate on kubelet, just to be on the safe side?

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+The `GenericInlineVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `InlineVolumeSource`,
```
s/InlineVolumeSource/EphemeralVolumeSource

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+The `GenericInlineVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `InlineVolumeSource`,
+- anything that specifically acts upon an `InlineVolumeSource` (scheduler,
```
s/InlineVolumeSource/EphemeralVolumeSource

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+### Feature gate
+
+The `GenericInlineVolumes` feature gate controls whether:
```
Should this be called GenericEphemeralVolume now?

pohly replied:
Okay. Do I now also need to rename the KEP to match?

xing-yang replied:
Yes

pohly replied:
I don't agree (see my other comment), but if that's what it takes to get the KEP merged... done.

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+  kubelet, etc.) instead of merely copying it (statefulset controller)
+  accepts the new volume source.
+
+Existing pods with such a volumes will not be scheduled respectively
```
s/such a volumes/`such volumes` or `such a volume`

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+### Modifying volumes
+
+Once the PVC for an inline volume has been created, it can be updated
```
s/inline/ephemeral

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+### Modifying volumes
+
+Once the PVC for an inline volume has been created, it can be updated
+directly like other PVCs. The controller will not interfere with that
```
Which controller?  Please clarify.

xing-yang replied:
s/The controller/The new controller

pohly replied:
Fixed, also in a few other places.

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+  well as negative case (feature disabled or feature used incorrectly).
+- A new [storage test
+  suite](https://github.com/kubernetes/kubernetes/blob/2b2cf8df303affd916bbeda8c2184b023f6ee53c/test/e2e/storage/testsuites/base.go#L84-L94)
+  will be added which tests inline volume creation, usage and deletion
```
s/inline/ephemeral

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
```
Which surveys are you talking about?

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+### Upgrade / Downgrade Strategy
+
+When downgrading to a cluster which does not support generic inline
```
s/inline/ephemeral

xing-yang replied:
This one is missed.

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+When downgrading to a cluster which does not support generic inline
+volumes (either by disabling the feature flag or an actual version
+downgrade), pods using generic inline volumes will no longer be
```
s/inline/ephemeral

xing-yang replied:
This one too.

pohly replied:
Updated. I consider the full name of the feature "generic ephemeral inline volumes". Both are relevant aspects of the volumes, so only mentioning one of them is arbitrary. "Generic" has to be added to distinguish from the existing "CSI ephemeral inline volumes".

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate
+    - Feature gate name: GenericInlineVolumes
```
GenericEphemeralVolumes

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+      - kubelet
+
+* **Does enabling the feature change any default behavior?**
+  If users are allowed to create pods but not PVCs, then generic inline volumes
```
s/inline/ephemeral

xing-yang replied:
This one is missed.

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
```
Remove sections that you have not filled out.

pohly replied:
Done.

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/kep.yaml:
```diff
@@ -0,0 +1,30 @@
+  beta: ""
+  stable: ""
+feature-gate:
+  name: GenericInlineVolumes
```
GenericEphemeralVolumes

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+  Avoid getting hung up on specific details and instead aim to get the goals of
+  the KEP clarified and merged quickly.  The best way to do this is to just
+  start with the high-level sections and fill out details incrementally in
+  subsequent PRs.
```
I wasn't sure which those are. The only safe one seems to be "Enhancement issue in release milestone, which links to KEP dir". I've marked that.

"Design details are appropriately documented" I think is okay, but perhaps someone else should confirm that?

"Test plan is in place" (yes), "giving consideration to SIG Architecture and SIG Testing input (no?)" - do they need to review?

"Graduation criteria is in place" - done.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+type EphemeralVolumeSource struct {
+    // Required. Defined as a pointer because in the future there
+    // might be alternative ways to specify how the ephemeral
+    // volume gets provisioned.
```
Currently it must not be nil. I've reworded a bit.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+type VolumeSource struct {
+   ...
+   CSI                   *CSIVolumeSource
+   PersistentVolumeClaim *PersistentVolumeClaimVolumeSource
```
I was assuming that readers are familiar with the existing API. Do I really need to explain it?

I've reworded the introduction a bit and added links.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+Recent releases of memcached added [support for using Persistent
+Memory](https://memcached.org/blog/persistent-memory/) (PMEM) instead
+of normal DRAM. When deploying memcached through one of the app
+controllers, `InlineVolumeSource` makes it possible to request a volume
```
Fixed, also below.

pohly commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,636 @@
+
+For those, it might make sense to mount the volume read-only inside
+the pod to prevent accidental modification of the data.
+
```
Added " For example, the goal might be to just retrieve the data and/or copy it elsewhere."

pohly commented:
I've updated the KEP and fixed kep.yaml. Please have another look.

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,565 @@
+
+* **Can the feature be disabled once it has been enabled (i.e. can we rollback
+  the enablement)?**
+  Yes, by disabling the feature gates. Existing pods with generic inline
```
This one is missed: s/inline/ephemeral

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
````diff
@@ -0,0 +1,565 @@
+```
+
+A new controller in `kube-controller-manager` is responsible for
+creating new PVCs for each such inline volume. It does that:
````
Here's another one: s/inline/ephemeral

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,565 @@
+## Example
+
+Here is a full example for a higher-level object that uses ephemeral
+inline volumes:
```
"inline" can be removed?

xing-yang commented on keps/sig-storage/1698-generic-inline-volumes/README.md:
```diff
@@ -0,0 +1,565 @@
+  Yes, by disabling the feature gates. Existing pods with generic inline
+  volumes that haven't started yet will not be able to start up anymore, because
+  kubelet does not know what to do with the volume. It also will not know how
+  to unmount generic inline volumes which would cause pods to get stuck, so nodes
```
s/inline/ephemeral

pohly commented:
@xing-yang I've renamed the KEP, please check again.

xing-yang commented on keps/sig-storage/1698-generic-ephemeral-volumes/kep.yaml:
```diff
@@ -0,0 +1,30 @@
+title: generic inline volumes
```
ephemeral

pohly replied:
Thanks, fixed.

xing-yang commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
````diff
@@ -0,0 +1,566 @@
+```
+
+A new controller in `kube-controller-manager` is responsible for
+creating new PVCs for each such inline volume. It does that:
````
ephemeral

pohly replied:
Fixed.

Let's hope this is the last one :smile:  :crossed_fingers:

pohly commented:
> Can you squash your commits?

Done, force-pushed.

kfox1111 commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+        hostPath:
+          path: /var/log
+      - name: scratch
+        ephemeral:
```
In general, this all looks good.

The only thing I'm not sure of is the naming here.

I would have loved to have the volume type ephemeral for the local inline ephemeral use case we have now. But I was told it had to hang off of the csi driver. 

Having some ephemeral be under csi and some under volumeSource would be hard to explain to users I think.

At some point, maybe pod api v2 whenever that happens, the csi driver should probably fold back out so it looks like its more at the root of the volumeSource struct. As all drivers but configmap/secret/emptydir/hostpath will be moving to csi, so having csi in the api kind of stops making sense at some point.

All in all, I'm not against this. I think in the long run its more usable. But, I think in the intermediate time its kind of confusing to end users. Perhaps we should design the api here such that it could be used for all kinds of ephemeral storage. And we treat the csi version as an alias and deprecate it? In my mind, the local ephemeral drivers are just an optimization (shortcuts) of the usage of general ephemeral inline. So to a user, ideally should look the same. They shouldn't have to care they are different.

pohly replied:
> But, I think in the intermediate time its kind of confusing to end users. Perhaps we should design the api here such that it could be used for all kinds of ephemeral storage. And we treat the csi version as an alias and deprecate it?

This is part of the long term plan and the reason why `VolumeClaimTemplate` is a pointer: we can add a new, alternative pointer to a struct with the values for a "CSI ephemeral inline volume", allow that as alternative and then with a suitable deprecation period, remove the `CSI` entry in `VolumeSource` in favor of this unified `Ephemeral` entry.

kfox1111 replied:
I can see how that might work with the raw api. but what about the yaml representation? How will it know what type of object to generate?

saad-ali replied:
Ya we might need a wrapper struct here `PersistentVolumeClaimTemplate` to disambiguate and be consistent with existing `PodTemplate`.

msau42 commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+Existing pods with such a volume will not be started when the feature
+gate is off.
+
+### Modifying volumes
```
What about updating the volume template in a higher level controller like deployment or statefulset? Would that cause a new rollout of the deployment/statefulset?

kfox1111 replied:
I would expect that to cause an update as its just a change from the deployment/statefulset/daemonset/etc's perspective. I think that's also a good behavior.

saad-ali replied:
Lifecycle should be tied to the pod. If higher level controller deletes pod and creates a new one, we should delete PVC and create a new one.

pohly replied:
At some point, the controller which created the pods has to delete the existing ones and create them anew, because updating the `Volumes` field in the pod spec is immutable (I checked). The new pods then start also with new PVCs/PVs.

I have not checked whether a deployment can update volumes in the replicaset.

msau42 commented on keps/sig-storage/1698-generic-ephemeral-volumes/kep.yaml:
```diff
@@ -0,0 +1,30 @@
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+status: provisional
```
Is this ready to be updated to implementable?

kfox1111 replied:
This to go in as an alpha next or beta?

pohly replied:
I hope so. Let me make the change, otherwise it will miss 1.19.

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+
+## Summary
+
+This KEP proposes an alternative mechanism for specifying volumes
+inside a pod spec. Those inline volumes then get converted to normal
```
nit: ...specifying **ephemeral** volumes inside a pod spec...

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+volume claim objects for provisioning, so storage drivers do not need
+to be modified.
+
+The lifecycle of those volumes is ephemeral: they will get deleted
+once the pod terminates.
```
nit: ...they will be created when the pod starts, and...

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+The lifecycle of those volumes is ephemeral: they will get deleted
+once the pod terminates.
+
+Because this is expected to be slower than [CSI ephemeral inline
```
nit: Clarify what slower means here (time to implementation, performance, etc.)?

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+- A normal, unmodified storage driver can be selected via a storage class.
+- The volume will be created using the normal storage provisioning
+  mechanism, without having to modify the driver or its deployment.
+- Storage capacity tracking can be enabled also for such volumes.
```
Eventual API alignment with CSI ephemeral inline volumes, as well as current ephemeral in-tree volumes (EmptyDir, etc.), to ensure consistency across ephemeral volume APIs and minimize user confusion.

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+    // Will be created as a stand-alone PVC to provision the volume.
+    // Required, must not be nil.
+    VolumeClaimTemplate *PersistentVolumeClaim
+    ReadOnly             bool
```
We should think twice about if this ReadOnly flag is really necessary (https://github.com/kubernetes/kubernetes/issues/70503). The answer may still be yes because it simplifies implementation, but if we can avoid making the problem worse, that would be great.

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+type EphemeralVolumeSource struct {
+    // Will be created as a stand-alone PVC to provision the volume.
+    // Required, must not be nil.
+    VolumeClaimTemplate *PersistentVolumeClaim
```
```suggestion
    VolumeClaimTemplate *PersistentVolumeClaimTemplate
```

Following the model of `PodTemplate`: https://github.com/kubernetes/kubernetes/blob/78f2c8f70d66dd5c64697418d034429b3ebd9dce/pkg/apis/core/types.go#L3143-L3150

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+concatenation of pod name and the `Volume.Name`. This is guaranteed to
+be unique for the pod and each volume in that pod. Care must be taken
+by the user to not exceed the length limit for object names, otherwise
+volumes cannot be created.
```
Maybe we should prevent setting any metadata. Arguably these PVCs will be an implementation detail not something the user should care about setting labels/annotations for? So maybe we only allow the user to set the spec in PVCTemplate?

saad-ali replied:
We can discuss this in the implementation PR.

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+new controller will accidentally modify unrelated PVCs.
+
+The volume scheduling library and kubelet must check that a PVC is
+owned by the pod before using it. Otherwise they must ignored it. This
```
How does it verify pod owns a PVC other then the name?

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+storage class for the PVC.
+
+Later, `kube-scheduler` and `external-provisioner` can be changed to
+automatically enable late binding for PVCs which are owned by a pod.
```
How will we do that? Just ignore the `VolumeBindingMode` in `StorageClass` if specified?

pohly replied:
Yes.

saad-ali replied:
Hmmm... pretty ugly wonder if we can do better. Can discuss during implementation.

saad-ali commented on keps/sig-storage/1698-generic-ephemeral-volumes/README.md:
```diff
@@ -0,0 +1,566 @@
+- Gather feedback from developers and surveys
+- Errors emitted as pod events
+- Decide whether `CSIVolumeSource` (in beta at the moment) should be
+  merged with `EphemeralVolumeSource`
```
* Decide whether in-tree ephemeral volume sources, like `EmptyDir` (GA already), should also be added `EphemeralVolumeSource` for sake of API consistency.

pohly replied:
Added

pohly commented:
> The diff is:
> ... 
>
> +The lifecycle of those volumes is ephemeral: they will get created
> +when the pod is created (for immediate [volume
> +binding](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode))
> +or when the pod starts (for `WaitForFirstConsumer` volume binding),
> +and deleted once the pod terminates.
> ...

That last change hadn't made it into the squashed commit and was still sitting in my tree - sorry! I've squashed it and force-pushed, now the commit is a080b55.

saad-ali commented:
Thanks @pohly 

/lgtm
/approve
